privacyidea/lib/challengeresponsedecorators.py

```diff
@@ -157,7 +160,44 @@ def generic_challenge_response_reset_pin(wrapped_function, *args, **kwds):
         # The tokenlist can contain more than one token. So we get the matching token object
         token_obj = next(t for t in args[0] if t.token.serial == serial)
         if g and Match.token(g, scope=SCOPE.AUTH, action=ACTION.CHANGE_PIN_VIA_VALIDATE, token_obj=token_obj).any():
-            reply_dict = _create_pin_reset_challenge(token_obj, _("Please enter a new PIN"))
+            reply_dict = _create_challenge(token_obj, CHALLENGE_TYPE.PIN_RESET, _("Please enter a new PIN"))
             return False, reply_dict
 
     return success, reply_dict
+
+
+def generic_challenge_response_resync(wrapped_function, *args, **kwds):
+    """
+    Check if the authentication request resulted an autosync
+
+    Conditions: To do so we check for "otp1c" in the tokeninfo data.
+
+    Policies: A policy defines, the token resync should be allowed this way.
+    Note: The general config "autoresync" needs to be set anyways.
+
+    args are:
+    :param tokenobject_list: The list of all the tokens of the user, that will be checked
+    :param passw: The password presented in the authentication.
+
+    kwds are:
+    :param options: options dictionary containing g
+    :param user: The user_obj
+    """
+    success, reply_dict = wrapped_function(*args, **kwds)
+
+    options = kwds.get("options") or {}
+    # After a failed authentication, we check if the token has an otp1c
+    if not success and reply_dict.get("serial"):
+        # FIXME: Only works if the one token has a unique PIN
+        serial = reply_dict.get("serial")
+        g = options.get("g")
+        # The tokenlist can contain more than one token. So we get the matching token object
+        token_obj = next(t for t in args[0] if t.token.serial == serial)
+        if token_obj.get_tokeninfo("otp1c"):
```

Comment:
Hm, if during the request the token is deleted, it could be possible that we do not find the token.

But we only end up at this code branch, if a token with the correct PIN identified by `reply_dict("serial")` could be found and does exist. So this token should exist.

Unless it would be deleted during the request or the serial would somehow somewhere get mangled.

But you are right, we could do:

~~~~python
if token_obj and token_obj.get_tokeninfo("otp1c"):
~~~~

Reply:
Will do.